### PR TITLE
Extend EMF validation

### PR DESCRIPTION
**Pull request by JeremyKuhne**
- Add support for save/restore DC
- Add support for world transforms
- Refactor to avoid duplication for DC state validation
- Add Poly16 validators
- Add context to XUnit exceptions to identify faulting validator
- Fix SkipAllValidator
- Start adding another string dump to look at applied transforms
- Add regression validation for DataGridView GridColor rendering

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/3890)

### CODE REVIEW

File: src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/EmfScope.cs [hunk 1]
```diff
@@ -136,6 +137,29 @@ bool stateTracker(ref EmfRecord record)
...
+                            case Gdi32.MWT.RIGHTMULTIPLY:
+                                state.Transform = state.Transform * transform->xform;
+                                break;
+                        }
+                        break;
+                    case Gdi32.EMR.SAVEDC:
+                        state.SaveDC();
+                        break;
+                    case Gdi32.EMR.RESTOREDC:
+                        break;
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/EmfScope.cs:162]: Are we missing `state.RestoreDc()` here? If not, perhaps it is worth adding a comment.

File: src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/RecordTypes/EMRSETCOLOR.cs [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 #nullable enable
 
+using System.Numerics;
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/RecordTypes/EMRSETCOLOR.cs:7]: [nit] Not needed, we can revert this file.

File: src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/EmfValidator.cs [hunk 3]
```diff
@@ -50,11 +61,19 @@ internal static class EmfValidator
             if (currentValidator != null)
             {
                 Assert.True(
-                    currentValidator.ValidationSatisfied,
+                    currentValidator.NeedsMoreRecords,
                     $"{currentValidator.GetType().Name} did not receive expected records");
             }
         }
 
+        private class WrappedXunitException : XunitException
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/EmfValidator.cs:69]: [nit] Please either move it to the end of the class ([guidelines](https://docs.microsoft.com/en-us/archive/blogs/brada/internal-coding-guidelines#28-file-organization)), or better yet - to own file.

File: src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/IEmfValidator.cs [hunk 4]
```diff
@@ -26,8 +26,8 @@ internal interface IEmfValidator
         void Validate(ref EmfRecord record, DeviceContextState state, out bool complete);
 
         /// <summary>
-        ///  Return true if this validator doesn't require any additional records.
+        ///  Return true if this validator doesn't require any additional records to satisfy it's validation.
         /// </summary>
-        bool ValidationSatisfied => false;
+        bool NeedsMoreRecords => false;
```
Comment (RussKie) [inline src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/IEmfValidator.cs:31]: I'm struggling to reconcile this - the name contradicts the comment and the value, e.g. 
> `NeedsMoreRecords == true` means "validator doesn't require any additional records to satisfy its validation"

The value must be inverted now, and the comment updated to reflect it. Also perhaps rename to `ExpectsMoreRecords`?

Comment (RussKie) [inline src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/IEmfValidator.cs:31]: 💭 This property value appears to be different only for `SkipAllValidator`, which pretty much make this property `SkipAllValidator`-specific...

**Review by RussKie — COMMENTED** [review]
🚀

File: src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/IEmfValidator.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/IEmfValidator.cs:31]: That is the only one that has it _currently_, but in any case it has to be part of the interface so I'm not sure what you're thinking here.

Comment (JeremyKuhne) [reply src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/IEmfValidator.cs:31]: I can't find a name I'm super satisfied with for these two "completion" booleans. I've inverted and changed the name to `FailIfIncomplete`.

**Comment by JeremyKuhne** [comment]
Debug build failed with #3898.

File: src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/IEmfValidator.cs (hunk 4 above)
Comment (RussKie) [reply src/System.Windows.Forms.Primitives/tests/TestUtilities/Metafiles/Validators/IEmfValidator.cs]: No specific thoughts or suggestion, just an observation.

**Review by RussKie — APPROVED** [review]

**Pull request merged**